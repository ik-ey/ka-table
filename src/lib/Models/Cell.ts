export class Cell {
    public columnKey!: string;
    public rowKeyValue!: any;
}
