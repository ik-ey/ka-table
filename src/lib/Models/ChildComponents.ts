import {
  ICellEditorProps,
  ICellProps,
  ICellTextProps,
  IDataRowProps,
  IEmptyCellProps,
  IFilterRowEditorProps,
  IGroupPanelCellProps,
  IGroupPanelProps,
  IGroupRowProps,
  IGroupSummaryCellProps,
  IGroupSummaryRowProps,
  IHeadCellProps,
  IHeadCellResizeProps,
  IHeadRowProps,
  IHeaderFilterButtonProps,
  ILoadingProps,
  INoDataRowProps,
  IPagingIndexProps,
  IPagingProps,
  IPagingSizeProps,
  IPopupContentItemProps,
  IPopupContentProps,
  IRowProps,
  ISortIconProps,
  ISummaryCellProps,
  ISummaryRowProps,
  ITableBodyProps,
  ITableFootProps,
  ITableHeadProps,
} from '../props';

import { ChildComponent } from './ChildComponent';
import { ITableProps } from '../';

export class ChildComponents {
  public cell?: ChildComponent<ICellProps>;
  public cellEditor?: ChildComponent<ICellEditorProps>;
  public cellEditorInput?: ChildComponent<ICellEditorProps>;
  public cellText?: ChildComponent<ICellTextProps>;
  public dataAndDetailsRows?: ChildComponent<IRowProps>;
  public dataRow?: ChildComponent<IDataRowProps>;
  public detailsCell?: ChildComponent<IDataRowProps>;
  public detailsRow?: ChildComponent<IDataRowProps>;
  public emptyCell?: ChildComponent<IEmptyCellProps>;
  public filterRowCell?: ChildComponent<IFilterRowEditorProps>;
  public groupPanel?: ChildComponent<IGroupPanelProps>;
  public groupPanelCell?: ChildComponent<IGroupPanelCellProps>;
  public groupExpandButton?: ChildComponent<IGroupRowProps>;
  public groupCell?: ChildComponent<IGroupRowProps>;
  public groupRow?: ChildComponent<IGroupRowProps>;
  public groupSummaryRow?: ChildComponent<IGroupSummaryRowProps>;
  public groupSummaryCell?: ChildComponent<IGroupSummaryCellProps>;
  public headFilterButton?: ChildComponent<IHeaderFilterButtonProps>;
  public headCell?: ChildComponent<IHeadCellProps>;
  public headCellContent?: ChildComponent<IHeadCellProps>;
  public headCellResize?: ChildComponent<IHeadCellResizeProps>;
  public headRow?: ChildComponent<IHeadRowProps>;
  public loading?: ChildComponent<ILoadingProps>;
  public noDataRow?: ChildComponent<INoDataRowProps>;
  public noDataCell?: ChildComponent<INoDataRowProps>;
  public paging?: ChildComponent<IPagingProps>;
  public pagingIndex?: ChildComponent<IPagingIndexProps>;
  public pagingPages?: ChildComponent<IPagingProps>;
  public pagingSize?: ChildComponent<IPagingSizeProps>;
  public pagingSizes?: ChildComponent<IPagingProps>;
  public popupContent?: ChildComponent<IPopupContentProps>;
  public popupContentItem?: ChildComponent<IPopupContentItemProps>;
  public rootDiv?: ChildComponent<ITableProps>;
  public sortIcon?: ChildComponent<ISortIconProps>;
  public summaryCell?: ChildComponent<ISummaryCellProps>;
  public summaryRow?: ChildComponent<ISummaryRowProps>;
  public table?: ChildComponent<ITableProps>;
  public tableBody?: ChildComponent<ITableBodyProps>;
  public tableFoot?: ChildComponent<ITableFootProps>;
  public tableHead?: ChildComponent<ITableHeadProps>;
  public tableWrapper?: ChildComponent<ITableProps>;
}
