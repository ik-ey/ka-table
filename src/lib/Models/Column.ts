import { DataType, SortDirection } from '../enums';

import { Field } from '../types';
import { PopupPosition } from './PopupPosition';

export class Column<TData = any> {
    public colGroup?: React.ColHTMLAttributes<HTMLElement>;
    public dataType?: DataType;
    public field?: Field;
    public filter?: (value: any, filterValue: any, rowData?: any) => boolean;
    public filterRowOperator?: any;
    public filterRowValue?: any;
    public headerFilterListItems?: (props: { data?: TData[] }) => string[];
    public headerFilterPopupPosition?: PopupPosition;
    public headerFilterSearch?: (value: any, searchValue: any, rowData?: any) => boolean;
    public headerFilterSearchValue?: any;
    public headerFilterValues?: string[];
    public isDraggable?: boolean;
    public isEditable?: boolean;
    public isFilterable?: boolean;
    public isHeaderFilterPopupShown?: boolean;
    public isHeaderFilterSearchable?: boolean;
    public isResizable?: boolean;
    public isSortable?: boolean;
    public key!: string;
    public sortDirection?: SortDirection;
    public sortIndex?: number;
    public style?: React.CSSProperties;
    public title?: string;
    public visible?: boolean;
    public width?: number | string;
}
