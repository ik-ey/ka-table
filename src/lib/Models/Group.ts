
export class Group {
    public columnKey!: string;
    public enableSummary?: boolean;
}
