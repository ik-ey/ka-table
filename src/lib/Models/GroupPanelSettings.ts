export class GroupPanelSettings {
    public enabled?: boolean;
    public text?: string;
    public deep?: number;
  }
