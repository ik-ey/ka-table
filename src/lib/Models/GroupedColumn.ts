
export class GroupedColumn {
    public key!: string;
    public title?: string;
    public columnsKeys!: string[];
}
