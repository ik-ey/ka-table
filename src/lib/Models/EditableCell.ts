import { Cell } from './Cell';

export class EditableCell extends Cell {
    public editorValue?: any;
    public validationMessage?: any;
}
