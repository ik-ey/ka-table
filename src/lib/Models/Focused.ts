import { Cell } from './Cell';

export class Focused {
    public cell?: Cell;
    public cellEditorInput?: Cell;
}
