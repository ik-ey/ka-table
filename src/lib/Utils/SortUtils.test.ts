import { DataType, SortDirection, SortingMode } from '../enums';
import {
  isMultipleSorting,
  isRemoteSorting,
  isSortingEnabled,
  isTripleStateSorting,
  sortColumns,
  sortData,
} from './SortUtils';

import { Column } from '../Models/Column';

const data: any[] = [
  { column: 1, id: 1 },
  { column: null, id: 6 },
  { column: 3, id: 2 },
  { column: 2, id: 3 },
  { column: 3, id: 4 },
  { column: 3, id: 5 },
  { column: null, id: 7 },
];

const columns: Column[] = [
  { key: 'id', title: 'Id', dataType: DataType.String },
  { key: 'column', title: 'Column 1', sortDirection: SortDirection.Descend, dataType: DataType.String },
];

describe('sortData', () => {
  describe('strings', () => {
    it('should be case insensitive', () => {
      const stringData: any[] = [
        { column: 'A', id: 1 },
        { column: null, id: 6 },
        { column: 'a', id: 2 },
        { column: 'Bb', id: 3 },
        { column: 'cC', id: 4 },
        { column: 'aa', id: 5 },
        { column: null, id: 7 },
      ];
      const newData = sortData(columns, stringData);
      expect(newData).toMatchSnapshot();
    });
  });
  it('should not change original data', () => {
    const newData = sortData(columns, data);
    expect(newData).not.toBe(data);
  });

  it('should be sorted by Descend', () => {
    const newData = sortData(columns, data);
    expect(newData).toMatchSnapshot();
  });

  it('should be sorted by Ascend', () => {
    const columns2 = [
      { key: 'column', title: 'Column 1', sortDirection: SortDirection.Ascend, dataType: DataType.String },
    ];
    const newData = sortData(columns2, data);
    expect(newData).toMatchSnapshot();
  });

  it('Custom logic', () => {
    const newData = sortData(columns, data, ({ column }) => {
      if (column.key === 'column') {
        return (a: any, b: any) => a === 3 ? -1 : 0;
      }
    });
    expect(newData).toMatchSnapshot();
  });
});

it('isTripleStateSorting', () => {
  expect(isTripleStateSorting(SortingMode.None)).toBeFalsy();
  expect(isTripleStateSorting(SortingMode.Single)).toBeFalsy();
  expect(isTripleStateSorting(SortingMode.SingleTripleState)).toBeTruthy();
  expect(isTripleStateSorting(SortingMode.SingleRemote)).toBeFalsy();
  expect(isTripleStateSorting(SortingMode.SingleTripleStateRemote)).toBeTruthy();
  expect(isTripleStateSorting(SortingMode.MultipleTripleStateRemote)).toBeTruthy();
});

it('isMultipleSorting', () => {
  expect(isMultipleSorting(SortingMode.None)).toBeFalsy();
  expect(isMultipleSorting(SortingMode.Single)).toBeFalsy();
  expect(isMultipleSorting(SortingMode.SingleTripleState)).toBeFalsy();
  expect(isMultipleSorting(SortingMode.SingleRemote)).toBeFalsy();
  expect(isMultipleSorting(SortingMode.SingleTripleStateRemote)).toBeFalsy();
  expect(isMultipleSorting(SortingMode.MultipleTripleStateRemote)).toBeTruthy();
});

it('isRemoteSorting', () => {
  expect(isRemoteSorting(SortingMode.None)).toBeFalsy();
  expect(isRemoteSorting(SortingMode.Single)).toBeFalsy();
  expect(isRemoteSorting(SortingMode.SingleTripleState)).toBeFalsy();
  expect(isRemoteSorting(SortingMode.SingleRemote)).toBeTruthy();
  expect(isRemoteSorting(SortingMode.SingleTripleStateRemote)).toBeTruthy();
  expect(isRemoteSorting(SortingMode.MultipleTripleStateRemote)).toBeTruthy();
});

it('isSortingEnabled', () => {
  expect(isSortingEnabled(SortingMode.None)).toBeFalsy();
  expect(isSortingEnabled(SortingMode.Single)).toBeTruthy();
  expect(isSortingEnabled(SortingMode.SingleTripleState)).toBeTruthy();
  expect(isSortingEnabled(SortingMode.SingleRemote)).toBeTruthy();
  expect(isSortingEnabled(SortingMode.SingleTripleStateRemote)).toBeTruthy();
  expect(isSortingEnabled(SortingMode.MultipleTripleStateRemote)).toBeTruthy();
});

it('isSortingEnabled column.isSortable', () => {
  expect(isSortingEnabled(SortingMode.Single, { isSortable: false } as Column)).toBeFalsy();
  expect(isSortingEnabled(SortingMode.Single, { isSortable: true } as Column)).toBeTruthy();
});


it('sortColumns', () => {
  const newData = sortColumns([{
    key: '1',
    sortDirection: SortDirection.Ascend,
    sortIndex: 1
  }, {
    key: '4',
    sortDirection: SortDirection.Descend,
  }, {
    key: '7',
    sortDirection: SortDirection.Descend,
  },  {
    key: '2',
    sortDirection: SortDirection.Descend,
    sortIndex: 2
  },  {
    key: '9',
    sortDirection: SortDirection.Descend,
    sortIndex: 9
  }, {
    key: '3',
    sortDirection: SortDirection.Descend,
    sortIndex: 3
  }, {
    key: '6',
  }, {
    key: '5',
  }]);
  expect(newData).toMatchSnapshot();
});
