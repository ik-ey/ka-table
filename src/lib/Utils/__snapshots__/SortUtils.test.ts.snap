// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sortColumns 1`] = `
Array [
  Object {
    "key": "4",
    "sortDirection": "descend",
  },
  Object {
    "key": "7",
    "sortDirection": "descend",
  },
  Object {
    "key": "1",
    "sortDirection": "ascend",
    "sortIndex": 1,
  },
  Object {
    "key": "2",
    "sortDirection": "descend",
    "sortIndex": 2,
  },
  Object {
    "key": "3",
    "sortDirection": "descend",
    "sortIndex": 3,
  },
  Object {
    "key": "9",
    "sortDirection": "descend",
    "sortIndex": 9,
  },
]
`;

exports[`sortData Custom logic 1`] = `
Array [
  Object {
    "column": 3,
    "id": 5,
  },
  Object {
    "column": 3,
    "id": 4,
  },
  Object {
    "column": 3,
    "id": 2,
  },
  Object {
    "column": 1,
    "id": 1,
  },
  Object {
    "column": null,
    "id": 6,
  },
  Object {
    "column": 2,
    "id": 3,
  },
  Object {
    "column": null,
    "id": 7,
  },
]
`;

exports[`sortData should be sorted by Ascend 1`] = `
Array [
  Object {
    "column": null,
    "id": 6,
  },
  Object {
    "column": null,
    "id": 7,
  },
  Object {
    "column": 1,
    "id": 1,
  },
  Object {
    "column": 2,
    "id": 3,
  },
  Object {
    "column": 3,
    "id": 2,
  },
  Object {
    "column": 3,
    "id": 4,
  },
  Object {
    "column": 3,
    "id": 5,
  },
]
`;

exports[`sortData should be sorted by Descend 1`] = `
Array [
  Object {
    "column": 3,
    "id": 2,
  },
  Object {
    "column": 3,
    "id": 4,
  },
  Object {
    "column": 3,
    "id": 5,
  },
  Object {
    "column": 2,
    "id": 3,
  },
  Object {
    "column": 1,
    "id": 1,
  },
  Object {
    "column": null,
    "id": 6,
  },
  Object {
    "column": null,
    "id": 7,
  },
]
`;

exports[`sortData strings should be case insensitive 1`] = `
Array [
  Object {
    "column": "cC",
    "id": 4,
  },
  Object {
    "column": "Bb",
    "id": 3,
  },
  Object {
    "column": "aa",
    "id": 5,
  },
  Object {
    "column": "A",
    "id": 1,
  },
  Object {
    "column": "a",
    "id": 2,
  },
  Object {
    "column": null,
    "id": 6,
  },
  Object {
    "column": null,
    "id": 7,
  },
]
`;
